### ISSUE TITLE
Visual Eval Pipeline: Phase 2 Comparison, Metrics & Orchestration

### ISSUE THREAD

**Issue by Trecek** [TalonT-Org/spectral-init#118]
## Summary

Extend the Phase 1 Python script with Phase 2 functionality: load Rust spectral init coordinates, run three-way SGD comparison (Python spectral vs Rust spectral vs random), produce comparison plots, compute quantitative metrics, and add an end-to-end orchestration script.

## Wave

**Wave 2** — depends on both:
- #116 (Phase 1 Baseline Script & Scaffolding)
- #117 (Rust Export Tool)

### 1. Phase 2 CLI

Extend `tests/visual_eval/generate_umap_comparisons.py` with `--phase compare`:

```bash
python tests/visual_eval/generate_umap_comparisons.py --phase compare [--dataset DATASET_NAME] [--output-dir DIR]
```

Phase 2 expects Phase 1 outputs AND Rust export outputs to exist in the output directory.

### 2. Three-Way SGD Comparison

For each dataset, run Python UMAP's SGD optimizer with three different initializations:

```python
# 1. Python spectral init → Python SGD (baseline)
mapper_py = umap.UMAP(init="spectral", random_state=42, n_neighbors=15, min_dist=0.1)
embed_py = mapper_py.fit_transform(X)

# 2. Rust spectral init → Python SGD (what we're testing)
rust_init = np.load(f"{name}_rust_init.npy")
mapper_rust = umap.UMAP(init=rust_init, random_state=42, n_neighbors=15, min_dist=0.1)
embed_rust = mapper_rust.fit_transform(X)

# 3. Random init → Python SGD (control — should look worse)
mapper_rand = umap.UMAP(init="random", random_state=42, n_neighbors=15, min_dist=0.1)
embed_rand = mapper_rand.fit_transform(X)
```

**Key**: UMAP's `init` parameter accepts any `(n, n_components)` numpy array. Our Rust output is in `[-10, 10]` with noise, which is exactly what UMAP's `noisy_scale_coords` produces, so it plugs in directly.

### 3. Three-Way Comparison Plot (2x3 grid)

Saved as `{name}_comparison.png`:

```
┌─────────────────────┬─────────────────────┬─────────────────────┐
│  Python Spectral    │  Rust Spectral      │  Random Init        │
│  Init (pre-SGD)     │  Init (pre-SGD)     │  (uniform random)   │
│  colored by label   │  colored by label   │  colored by label   │
├─────────────────────┼─────────────────────┼─────────────────────┤
│  Python Init →      │  Rust Init →        │  Random Init →      │
│  Python SGD         │  Python SGD         │  Python SGD         │
│  (final embedding)  │  (final embedding)  │  (final embedding)  │
└─────────────────────┴─────────────────────┴─────────────────────┘
```

- Top row: pre-SGD initialization coordinates
- **Bottom row is the key comparison**: if Python→SGD and Rust→SGD look the same, our spectral init is correct
- For the random init pre-SGD panel: show the random initial coordinates (uniform in [-10, 10])

**Plot style**: Same as Phase 1 (point size 1-3, alpha 0.5, tab10/tab20 colormap, no axis labels, dataset+strategy in title).

### 4. Overlay Plot (1x1)

Saved as `{name}_overlay.png`:

Overlay Python SGD and Rust SGD embeddings on the same axes:
- Python SGD points in blue, marker='o', alpha=0.3
- Rust SGD points in red, marker='x', alpha=0.3
- Legend identifying each
- Title: `"{name}: Python vs Rust SGD Overlay"`

Points should be nearly co-located if the inits are equivalent.

### 5. Quantitative Metrics

Compute per-dataset metrics comparing all three strategies:

| Metric | Python Init → SGD | Rust Init → SGD | Random Init → SGD |
|---|---|---|---|
| Trustworthiness (k=15) | `sklearn.manifold.trustworthiness(X, embed, n_neighbors=15)` | ... | ... |
| Silhouette score | `sklearn.metrics.silhouette_score(embed, labels)` | ... | ... |
| Procrustes distance (vs Python) | — | `scipy.spatial.procrustes(embed_py, embed_rust)[2]` | `scipy.spatial.procrustes(embed_py, embed_rand)[2]` |
| Pairwise distance correlation | — | Pearson corr of all pairwise Euclidean distances | ... |

**Pairwise distance correlation**: Compute all `n*(n-1)/2` pairwise Euclidean distances in each embedding, then Pearson correlation between the Python and Rust distance vectors. For large n, subsample to ~2000 points.

#### Success Criteria Thresholds

Report PASS/FAIL for Rust vs Python:

| Metric | Threshold | Interpretation |
|---|---|---|
| Procrustes distance | < 0.05 | Embeddings geometrically equivalent |
| Pairwise distance correlation | > 0.99 | Global structure preserved |
| Trustworthiness difference | < 0.01 | Local structure equivalent |
| Silhouette score difference | < 0.05 | Cluster quality equivalent |

If Random Init → SGD also passes these thresholds, print a warning: "Dataset too easy — spectral init provides no advantage over random."

### 6. Metrics JSON Export

Save per-dataset metrics as `{name}_metrics.json`:

```json
{
    "dataset": "blobs_1000",
    "n_samples": 1000,
    "n_features": 10,
    "python_spectral": {
        "trustworthiness": 0.98,
        "silhouette": 0.72
    },
    "rust_spectral": {
        "trustworthiness": 0.98,
        "silhouette": 0.71,
        "procrustes_vs_python": 0.012,
        "pairwise_corr_vs_python": 0.997
    },
    "random": {
        "trustworthiness": 0.95,
        "silhouette": 0.65,
        "procrustes_vs_python": 0.34,
        "pairwise_corr_vs_python": 0.82
    },
    "pass_fail": {
        "procrustes": "PASS",
        "pairwise_corr": "PASS",
        "trustworthiness": "PASS",
        "silhouette": "PASS",
        "overall": "PASS"
    }
}
```

### 7. Orchestration Script

Create `tests/visual_eval/run_eval.sh`:

```bash
#!/usr/bin/env bash
set -euo pipefail

SCRIPT_DIR="$(cd "$(dirname "$0")" && pwd)"
PROJECT_ROOT="$(cd "$SCRIPT_DIR/../.." && pwd)"

echo "=== Phase 1: Generate baselines ==="
python "$SCRIPT_DIR/generate_umap_comparisons.py" --phase baseline

echo ""
echo "=== Rust Export: Generate spectral init coordinates ==="
cd "$PROJECT_ROOT"
cargo test --test export_rust_init --features testing -- --ignored --nocapture

echo ""
echo "=== Phase 2: Compare and evaluate ==="
python "$SCRIPT_DIR/generate_umap_comparisons.py" --phase compare

echo ""
echo "=== Summary ==="
# Print pass/fail from metrics JSONs
python -c "
import json, glob
for f in sorted(glob.glob('$SCRIPT_DIR/output/*_metrics.json')):
    m = json.load(open(f))
    status = m['pass_fail']['overall']
    print(f\"  {m['dataset']:25s} {status}\")
"
```

Make it executable (`chmod +x`).

### 8. Update README

Update `tests/visual_eval/README.md` (created in #116) to document:
- Phase 2 workflow
- Orchestration script usage (`./tests/visual_eval/run_eval.sh`)
- Success criteria and what PASS/FAIL means
- Example output

## Implementation Notes

- For Procrustes: `scipy.spatial.procrustes` expects matrices of same shape, returns `(mtx1, mtx2, disparity)` where `disparity` is the Procrustes distance
- For pairwise distance correlation on large datasets (n > 2000): subsample to avoid O(n²) memory. Use `np.random.RandomState(42).choice(n, 2000, replace=False)` for deterministic subsampling
- The Python spectral init pre-SGD coordinates were already saved in Phase 1 as `{name}_py_spectral.npy` — load these for the comparison top-left panel
- The Rust pre-SGD coordinates are `{name}_rust_init.npy` — load for top-center panel
- For random init pre-SGD: generate with `np.random.RandomState(42).uniform(-10, 10, (n, 2))` to show what random looks like
- Phase 2 needs to re-fit UMAP to get the SGD embeddings. For the Python spectral case, the full SGD result was already saved as `{name}_py_final.npy` — but the Rust→SGD and random→SGD runs are new
- All UMAP parameters must be identical across the three strategies (same n_neighbors, min_dist, random_state)

## Acceptance Criteria

- [ ] `--phase compare` works and produces comparison plots for all datasets
- [ ] 2x3 comparison plots saved as `{name}_comparison.png`
- [ ] Overlay plots saved as `{name}_overlay.png`
- [ ] Metrics JSON saved as `{name}_metrics.json` with all fields
- [ ] Pass/fail assessment printed to stdout
- [ ] `tests/visual_eval/run_eval.sh` orchestrates the full pipeline end-to-end
- [ ] README updated with Phase 2 documentation
- [ ] Script gracefully handles missing Rust init files (skip dataset with warning)

## Files Created/Modified

- **Modify:** `tests/visual_eval/generate_umap_comparisons.py` (add Phase 2 functionality)
- **Create:** `tests/visual_eval/run_eval.sh` (orchestration script)
- **Modify:** `tests/visual_eval/README.md` (add Phase 2 docs)
